server/aesdsocket.c

```diff
@@ -53,39 +78,183 @@ const char* errs[] = {
 		xstat = -1;
 	}
 	closelog();
 
 	// clean up state
-	if (csock != -1) close(csock);
+	struct node* p;
+	for (struct node* n = thread_ll; n; n = p) {
+		p = n->next;
+	   	free(n);
+	}	
 	if (asock != -1) close(asock);
 	if (of_mem) munmap(of_mem, of_memsz);
 	if (ofd != -1) close (ofd);
-	if (packet) free(packet);
 	unlink(outpath);
 
 	exit(xstat);
 }
 
-// wrapper to conform to sigaction struct def
+// synchronous main thread cleanup function
+__attribute__((noreturn))
+static void cleanup (cleanup_src src) {
+	cleanup_errno(src, errno);
+}
+
+// async SIGTERM/SIGINT handler
 __attribute__((noreturn))
-static void sigcleanup (int sig) {
+static void sigcleanup (int sig, siginfo_t* info, void* unused) {
 	assert(sig == SIGTERM || sig == SIGINT);
-	cleanup((cleanup_src) sig);
+	if (info->si_code == SI_QUEUE) {
+		int32_t* unpack = (int32_t*) &info->si_ptr; 
+		cleanup_errno((cleanup_src) unpack[0], unpack[1]);
+	} else cleanup_errno((cleanup_src) sig, 0);
+}
+
+// cleanup handler for threads (sends SIGTERM with error info)
+__attribute__((noreturn))
+static void cleanup_thr (cleanup_src src) {
+	int32_t pack[2] = {src, errno};
+	union sigval v = {.sival_ptr = (void*) *(uint64_t*) pack};
+	sigqueue(getpid(), SIGTERM, v);
+	pthread_exit(NULL);
+}
+
+//-----client thread-----
+
+// close socket if cancelled
+static void vclose(void* fdpt) {
+	close(*(int*) fdpt);
+}
+
+static void* client_thread (void* param_v) {
+	// these must be declared before pushing the cleanup handler
+	// because it creates a scope...
+	int s;
+	char* cip;
+	struct client* param = (struct client*) param_v;
+	pthread_cleanup_push(vclose, &param->sock);
+
+	// log accepted connection
+	cip = alloca(INET_ADDRSTRLEN);
+	pthread_mutex_lock(&ntoa_lk);
+	const char* ipstr = inet_ntoa(param->addr);
+	strncpy(cip, ipstr, INET_ADDRSTRLEN);
+	pthread_mutex_unlock(&ntoa_lk);
+	syslog(LOG_INFO, "Accepted connection from %s\n", cip);
+	
+	// alloc packet buffer
+	size_t max_packet = INITIAL_MAX_PACKET;
+	char* packet = malloc(max_packet);
+	if (!packet) cleanup_thr(SRC_MALLOC);
+	pthread_cleanup_push(free, packet);
+
+	// get packet
+	char* delim = NULL;
+	size_t packet_sz = 0;
+	size_t max_read = max_packet;
+	while (!delim) {
+		assert(packet_sz <= max_packet);
+		ssize_t read_sz = read(param->sock, &packet[packet_sz], max_read);	
+		if (read_sz == -1)
+			cleanup_thr(SRC_READ);
+		
+		if (read_sz == max_read) { // need to expand buffer
+			max_read = max_packet;
+			max_packet *= 2;
+			packet = realloc(packet, max_packet);
+			if (!packet) cleanup_thr(SRC_REALLOC);
+		}
+
+		delim = memchr(&packet[packet_sz], '\n', read_sz);
+		packet_sz += read_sz;
+		max_read -= read_sz;
+	}
+
+	// write packet
+	pthread_mutex_lock(&of_lk);
+	if (of_sz + packet_sz > of_memsz) { // need to expand mapping
+		void* prev_mem = of_mem;
+		size_t new_memsz = ((of_sz + packet_sz)/PAGE_SIZE + 1)*PAGE_SIZE;
+		s = ftruncate(ofd, new_memsz);
+		if (s == -1) cleanup_thr(SRC_FTRUNCATE);
```

Comment:
@georgehodgkins what's going to happen to your mutex if your cleanup_thread handler occurs while it's held?